src/lib/config.ts

```diff
@@ -0,0 +1,417 @@
+    );
+  }
+}
+
+// Represents `~/.fauna-shell`
+class RootConfig {
+  defaultEndpoint?: string;
+  endpoints: { [key: string]: Endpoint };
+
+  constructor(config: Config) {
+    this.defaultEndpoint = config.strOpt("default");
+    this.endpoints = Object.fromEntries(
+      config
+        .entries((k) => k !== "default")
+        .map(([k, v]) => [k, new Endpoint(v)])
+    );
+
+    if (this.defaultEndpoint === "default") {
+      throw new InvalidConfigError(
+        "Default endpoint cannot be named 'default'"
+      );
+    } else if (
+      this.defaultEndpoint &&
+      this.endpoints[this.defaultEndpoint] === undefined
+    ) {
+      throw new InvalidConfigError(
+        `Default endpoint '${this.defaultEndpoint}' was not found`
+      );
+    }
+  }
+}
+
+// Represents `.fauna-project`
+class LocalConfig {
+  defaultStack: string | undefined;
+  stacks: { [key: string]: Stack };
+
+  constructor(config: Config) {
+    this.defaultStack = config.strOpt("default");
+    this.stacks = Object.fromEntries(
+      config.entries((k) => k !== "default").map(([k, v]) => [k, new Stack(v)])
+    );
+
+    if (this.defaultStack === "default") {
+      throw new InvalidConfigError("Default stack cannot be named 'default'");
+    } else if (
+      this.defaultStack &&
+      this.stacks[this.defaultStack] === undefined
+    ) {
+      throw new InvalidConfigError(
+        `Default stack '${this.defaultStack}' was not found`
+      );
+    }
+  }
+
+  validate(rootConfig: RootConfig) {
+    for (const stack of Object.values(this.stacks)) {
+      if (rootConfig.endpoints[stack.endpoint] === undefined) {
+        throw new InvalidConfigError(
+          `Endpoint '${stack.endpoint}' not found in ~/.fauna-shell`
+        );
+      }
+    }
+  }
+}
+
+class Endpoint {
+  secret: string;
+  url?: string;
+
+  graphqlHost?: string;
+  graphqlPort?: number;
+
+  constructor(config: Config) {
+    this.secret = config.str("secret");
+    this.url = getURLFromConfig(config);
+
+    this.graphqlHost = config.strOpt("graphqlHost");
+    this.graphqlPort = config.numberOpt("graphqlPort");
+  }
+
+  makeEndpoint(
+    scope?: string,
+    role?: string
+  ): {
+    secret: string;
+    url: string | undefined;
+    graphqlHost: string | undefined;
+    graphqlPort: number | undefined;
+  } {
+    const secret =
+      this.secret + (scope ? `:${scope}` : "") + ":" + (role ?? "admin");
+
+    return {
+      secret,
+      url: this.url,
+      graphqlHost: this.graphqlHost,
+      graphqlPort: this.graphqlPort,
+    };
+  }
+}
+
+class Stack {
+  /**
+   * The endpoint name to use as a base.
+   */
+  endpoint: string;
+  /**
+   * The database path to use.
+   */
+  database: string | undefined;
+
+  constructor(config: Config) {
+    this.endpoint = config.str("endpoint");
+    this.database = config.strOpt("database");
+  }
+
+  updateDatabase(database: string | undefined) {
+    if (database !== undefined) {
```

Comment:
what if I'm trying to unset the database?

Reply:
This function isn't used. The `database` field is public, so it should just be assigned to.